src/feature/NodeGraphFeature.ts

```diff
@@ -0,0 +1,183 @@
+  private shownLanguageServerNotAvailable = false;
+
+  constructor(connectionManager:IConnectionManager) {
+    this.connectionManager = connectionManager;
+  }
+
+  public provideTextDocumentContent(uri: vscode.Uri): Thenable<string> {
+    const sourceUri = vscode.Uri.parse(uri.query);
+
+    return vscode.workspace.openTextDocument(sourceUri).then(document => {
+      const initialData = {
+        previewUri: uri.toString(),
+        source: sourceUri.toString()
+      };
+
+      if ((this.connectionManager.status !== ConnectionStatus.Running) && (this.connectionManager.status !== ConnectionStatus.Starting)) {
+        if (this.shownLanguageServerNotAvailable) {
+          vscode.window.showInformationMessage("The Puppet Node Graph Preview is not available as the Editor Service is not ready");
+          this.shownLanguageServerNotAvailable = true;
+        }
+        return "The Puppet Node Graph Preview is not available as the Editor Service is not ready";
+      }
+
+      // Use the language server to render the document
+      return this.connectionManager.languageClient
+        .sendRequest(CompileNodeGraphRequest.type, sourceUri)
+        .then(
+          (compileResult) => {
+
+          var svgContent = '';
+          if (compileResult.dotContent !== null) {
+            var styling = `
+            bgcolor = "transparent"
+            color = "white"
+            rankdir = "TB"
+            node [ shape="box" penwidth="2" color="#e0e0e0" style="rounded,filled" fontname="Courier New" fillcolor=black, fontcolor="white"]
+            edge [ style="bold" color="#f0f0f0" penwith="2" ]
+
+            label = ""`
+
+            var graphContent = compileResult.dotContent;
+            // vis.jz sees backslashes as escape characters, however they are not in the DOT language.  Instead
+            // we should escape any backslash coming from a valid DOT file in preparation to be rendered
+            graphContent = graphContent.replace(/\\/g,"\\\\");
+            graphContent = graphContent.replace(`label = "vscode"`,styling);
+
+            svgContent = viz(graphContent,"svg");
+          }
+
+          var errorContent = `<div style='font-size: 1.5em'>${compileResult.error}</div>`
+          if (compileResult.error === null) { errorContent = ''; }
+
+          if (reporter) {
+            reporter.sendTelemetryEvent(PuppetNodeGraphToTheSideCommandId);
+          }
+
+          return `
+            ${errorContent}
+            <div id="graphviz_svg_div">
+              ${svgContent}
+            </div>`;
+      });
+    });
+  }
+
+  get onDidChange(): vscode.Event<vscode.Uri> {
+    return this.onDidChangeEvent.event;
+  }
+
+  public update(uri: vscode.Uri) {
+    if (!this.waiting) {
+      this.waiting = true;
+      setTimeout(() => {
+        this.waiting = false;
+        this.onDidChangeEvent.fire(uri);
+      }, 300);
+    }
+  }
+}
+
+export class NodeGraphFeature implements IFeature {
```

Comment:
Is it a typescript thing to put multiple classes in one file? While this is all one 'feature', does it need to be all in one file?

Reply:
I forgot to remove the export from the provider.  From a TS point of view, only the feature class is exported.  so one class - one file.